File: services/auth-service/telemetry/telemetry.go
// Package telemetry — inicialização do OpenTelemetry para serviços Go
// ===================================================================
//
// Este pacote provê uma função `Init` que:
//   1. Configura um TracerProvider apontando para o OTel Collector via OTLP/HTTP
//   2. Configura um MeterProvider que envia métricas via OTLP E expõe
//      /metrics em :9464 (formato Prometheus) — backup para o caso do
//      OTel Collector cair.
//   3. Retorna um http.Handler que é um WRAPPER do mux original adicionando
//      spans automáticos a cada request (otelhttp middleware).
//
// Por que separar do main.go?
//   - O auth-service e o evaluation-service compartilham EXATAMENTE a mesma
//     lógica de bootstrap. Em vez de copiar 80 linhas no main.go de cada
//     um, isolamos aqui. Cada serviço importa "./telemetry" do seu próprio
//     contexto de build.
//
// Variáveis de ambiente respeitadas:
//   OTEL_SERVICE_NAME            — nome do serviço (ex: auth-service)
//   OTEL_EXPORTER_OTLP_ENDPOINT  — endpoint do coletor (default OTLP http)
//   DISABLE_OTEL=true            — desliga TODA a instrumentação (debug local)
package telemetry

import (
	"context"
	"log"
	"net/http"
	"os"
	"strconv"
	"time"

	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/exporters/prometheus"
	metricapi "go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

// Shutdown é a função retornada por Init para flush final dos buffers.
// Chame-a em defer no main(). Sem isso, spans em buffer NÃO chegam ao
// coletor quando o processo termina (kill, SIGTERM no rolling update).
type Shutdown func(context.Context) error

// noopShutdown é o que retornamos quando OTel está desligado.
func noopShutdown(context.Context) error { return nil }

// Init prepara o OTel para um serviço Go.
//
// Retornos:
//   - shutdown: função de cleanup
//   - error:    se a inicialização falhar (não fatal — ver chamadas no main)
func Init(ctx context.Context, serviceName string) (Shutdown, error) {
	if os.Getenv("DISABLE_OTEL") == "true" {
		log.Println("OpenTelemetry desabilitado via DISABLE_OTEL=true")
		return noopShutdown, nil
	}

	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		endpoint = "otel-opentelemetry-collector.observability.svc.cluster.local:4318"
	}
	// As libs OTel esperam "host:port" SEM esquema. Tiramos se vier por engano.
	endpoint = stripScheme(endpoint)

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
			semconv.ServiceVersion(envOr("SERVICE_VERSION", "1.0.0")),
			semconv.DeploymentEnvironment(envOr("DEPLOYMENT_ENV", "production")),
			semconv.ServiceNamespace("togglemaster"),
		),
		resource.WithFromEnv(),
		resource.WithProcess(),
		resource.WithHost(),
	)
	if err != nil {
		return noopShutdown, err
	}

	// -------- Traces --------
	traceExp, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(endpoint),
		otlptracehttp.WithInsecure(), // dentro do cluster, mTLS não é necessário
	)
	if err != nil {
		return noopShutdown, err
	}
	tracerProvider := sdktrace.NewTracerProvider(
		sdktrace.WithResource(res),
		sdktrace.WithBatcher(traceExp,
			sdktrace.WithBatchTimeout(5*time.Second),
			sdktrace.WithMaxExportBatchSize(512),
		),
	)
	otel.SetTracerProvider(tracerProvider)

	// Propagator: W3C TraceContext (padrão de mercado) + Baggage
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	// -------- Metrics --------
	// 2 leitores:
	//  a) OTLP -> Collector -> Prometheus/Datadog
	//  b) /metrics local (Prometheus scrape direto, fallback)
	metricExp, err := otlpmetrichttp.New(ctx,
		otlpmetrichttp.WithEndpoint(endpoint),
		otlpmetrichttp.WithInsecure(),
	)
	if err != nil {
		return noopShutdown, err
	}
	promReader, err := prometheus.New()
	if err != nil {
		return noopShutdown, err
	}
	meterProvider := metric.NewMeterProvider(
		metric.WithResource(res),
		metric.WithReader(promReader),
		metric.WithReader(metric.NewPeriodicReader(metricExp,
			metric.WithInterval(30*time.Second),
		)),
	)
	otel.SetMeterProvider(meterProvider)

	// Sobe o /metrics em :9464 numa goroutine (parecido com o prometheus-client)
	go func() {
		mux := http.NewServeMux()
		mux.Handle("/metrics", promhttpHandler())
		s := &http.Server{
			Addr:              ":9464",
			Handler:           mux,
			ReadHeaderTimeout: 5 * time.Second,
		}
		log.Println("Prometheus /metrics escutando em :9464")
		if err := s.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Printf("Erro no servidor de métricas: %v", err)
		}
	}()

	log.Printf("OpenTelemetry inicializado para %q (endpoint=%s)", serviceName, endpoint)

	shutdown := func(ctx context.Context) error {
		// Importante: chamar AMBOS para fazer flush final
		_ = tracerProvider.Shutdown(ctx)
		_ = meterProvider.Shutdown(ctx)
		return nil
	}
	return shutdown, nil
}

// WrapHandler embrulha um http.Handler com:
//  1) o middleware otelhttp (gera spans para o tracing distribuído / Service Map)
//  2) um middleware de MÉTRICAS CUSTOMIZADAS com nomes DETERMINÍSTICOS
//     (http_requests_total e http_request_duration_seconds), idênticos aos
//     emitidos pelos serviços Python. Sem isso, o otelhttp emitiria
//     `http.server.request.duration` cuja tradução para Prometheus é
//     imprevisível entre versões, e os alertas/dashboard ficariam vazios.
func WrapHandler(handler http.Handler, serverName string) http.Handler {
	meter := otel.GetMeterProvider().Meter("togglemaster.http")

	// Counter: Prometheus adiciona o sufixo `_total` -> http_requests_total
	requestsTotal, _ := meter.Int64Counter(
		"http_requests",
		metricapi.WithDescription("Total de requisições HTTP processadas"),
		metricapi.WithUnit("1"),
	)
	// Histogram em segundos, com buckets explícitos adequados a uma API web
	requestDuration, _ := meter.Float64Histogram(
		"http_request_duration_seconds",
		metricapi.WithDescription("Duração das requisições HTTP em segundos"),
		metricapi.WithUnit("s"),
		metricapi.WithExplicitBucketBoundaries(
			0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5,
			0.75, 1.0, 2.0, 5.0, 10.0,
		),
	)

	serviceName := envOr("OTEL_SERVICE_NAME", serverName)

	metricsMiddleware := func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()
			// Captura o status code via ResponseWriter wrapper
			rw := &statusRecorder{ResponseWriter: w, status: http.StatusOK}
			next.ServeHTTP(rw, r)
			elapsed := time.Since(start).Seconds()

			attrs := metricapi.WithAttributes(
				attribute.String("service", serviceName),
				attribute.String("http_request_method", r.Method),
				attribute.String("http_route", r.URL.Path),
			)
			statusAttrs := metricapi.WithAttributes(
				attribute.String("service", serviceName),
				attribute.String("http_request_method", r.Method),
				attribute.String("http_route", r.URL.Path),
				attribute.String("http_response_status_code", strconv.Itoa(rw.status)),
			)
			requestsTotal.Add(r.Context(), 1, statusAttrs)
			requestDuration.Record(r.Context(), elapsed, attrs)
		})
	}

	// Ordem: otelhttp por fora (cria o span raiz), métricas por dentro.
	return otelhttp.NewHandler(metricsMiddleware(handler), serverName)
}

// statusRecorder captura o status code escrito pelo handler, para podermos
// rotular a métrica http_requests_total com http_response_status_code.
type statusRecorder struct {
	http.ResponseWriter
	status int
}

func (r *statusRecorder) WriteHeader(code int) {
	r.status = code
	r.ResponseWriter.WriteHeader(code)
}

// WrapTransport embrulha o http.RoundTripper para que chamadas HTTP DE SAÍDA
// propaguem o trace context (header `traceparent` do W3C) para o próximo
// serviço. É a metade "client" do tracing distribuído — sem isso, cada
// serviço cria seu próprio trace isolado e o Service Map fica desconexo.
func WrapTransport(base http.RoundTripper) http.RoundTripper {
	if base == nil {
		base = http.DefaultTransport
	}
	return otelhttp.NewTransport(base)
}

// --- helpers ---

func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func stripScheme(s string) string {
	for _, p := range []string{"http://", "https://"} {
		if len(s) > len(p) && s[:len(p)] == p {
			return s[len(p):]
		}
	}
	return s
}

File: services/flag-service/telemetry.py
"""
ToggleMaster — Módulo compartilhado de instrumentação OpenTelemetry
====================================================================

Por que um módulo compartilhado em vez de copiar para cada serviço?
- Os 3 serviços Python (flag, targeting, analytics) precisam EXATAMENTE da
  mesma inicialização. Copiar 70 linhas em 3 lugares = 3 lugares para errar.
- Cada serviço importa este módulo no início do seu app.py com
  `from telemetry import init_telemetry`.

O que este módulo faz (em ordem):
1. Lê variáveis de ambiente (OTEL_SERVICE_NAME, OTEL_EXPORTER_OTLP_ENDPOINT)
2. Configura o TracerProvider e MeterProvider apontando para o OTel Collector
3. Auto-instrumenta Flask, Requests, Psycopg2, Botocore (não precisamos
   decorar HANDLER POR HANDLER — a auto-instrumentação injeta spans em
   TODA chamada das libs).
4. Inicia o servidor de métricas Prometheus na porta 9464 (formato compatível
   com `prometheus.io/scrape: true`).

IMPORTANTE: as funções `instrument_*` são idempotentes E podem ser chamadas
SOMENTE depois de criar `app = Flask(__name__)`. Por isso o init_telemetry
recebe o app como parâmetro.
"""

import logging
import os

log = logging.getLogger(__name__)


def init_telemetry(flask_app=None, service_name: str | None = None) -> None:
    """
    Inicializa OpenTelemetry para o microsserviço.

    Parâmetros:
        flask_app:    instância do Flask (ou None se não for um serviço HTTP)
        service_name: nome do serviço; se omitido lê de OTEL_SERVICE_NAME

    Se a env var DISABLE_OTEL=true estiver setada, NADA é feito (útil em
    rodadas locais sem o coletor).
    """
    if os.getenv("DISABLE_OTEL", "false").lower() == "true":
        log.info("OpenTelemetry desabilitado via DISABLE_OTEL=true")
        return

    service_name = service_name or os.getenv("OTEL_SERVICE_NAME", "unknown-service")
    endpoint = os.getenv(
        "OTEL_EXPORTER_OTLP_ENDPOINT",
        "http://otel-opentelemetry-collector.observability.svc.cluster.local:4318",
    )

    try:
        # --- 1. Resource attributes (vão em TODO span/metric/log) ---
        from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION

        resource = Resource.create({
            SERVICE_NAME: service_name,
            SERVICE_VERSION: os.getenv("SERVICE_VERSION", "1.0.0"),
            "deployment.environment": os.getenv("DEPLOYMENT_ENV", "production"),
            "service.namespace": "togglemaster",
        })

        # --- 2. TracerProvider (traces) ---
        from opentelemetry import trace
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
            OTLPSpanExporter,
        )

        provider = TracerProvider(resource=resource)
        provider.add_span_processor(
            BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{endpoint}/v1/traces"))
        )
        trace.set_tracer_provider(provider)

        # --- 3. MeterProvider (métricas) ---
        # 2 leitores: 1 que exporta para o Collector via OTLP E outro que
        # expõe /metrics no formato Prometheus na porta 9464. Por que os 2?
        # - OTLP: roteia para Prometheus + Datadog (caminho oficial OTel)
        # - /metrics: backup que funciona MESMO se o coletor cair. Util para
        #   o Prometheus continuar scrapando direto em uma falha do OTel.
        from opentelemetry import metrics
        from opentelemetry.sdk.metrics import MeterProvider
        from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
        from opentelemetry.exporter.otlp.proto.http.metric_exporter import (
            OTLPMetricExporter,
        )
        from opentelemetry.exporter.prometheus import PrometheusMetricReader
        from prometheus_client import start_http_server

        # /metrics na porta 9464 (default do PrometheusMetricReader)
        prom_reader = PrometheusMetricReader()
        start_http_server(port=9464, addr="0.0.0.0")

        # OTLP para o coletor a cada 30s
        otlp_reader = PeriodicExportingMetricReader(
            OTLPMetricExporter(endpoint=f"{endpoint}/v1/metrics"),
            export_interval_millis=30_000,
        )

        # IMPORTANTE: os buckets DEFAULT do histogram OTel são pensados para
        # MILISSEGUNDOS (0, 5, 10, ..., 10000). Como medimos em SEGUNDOS, sem
        # uma View os buckets ficariam errados e o histogram_quantile do
        # alerta HighLatency (> 2s) daria valores sem sentido. A View abaixo
        # define explicitamente fronteiras em segundos adequadas a uma API web.
        from opentelemetry.sdk.metrics.view import View, ExplicitBucketHistogramAggregation

        latency_view = View(
            instrument_name="http_request_duration_seconds",
            aggregation=ExplicitBucketHistogramAggregation(
                boundaries=[
                    0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5,
                    0.75, 1.0, 2.0, 5.0, 10.0,
                ]
            ),
        )

        meter_provider = MeterProvider(
            resource=resource,
            metric_readers=[prom_reader, otlp_reader],
            views=[latency_view],
        )
        metrics.set_meter_provider(meter_provider)

        # --- 4. Auto-instrumentação ---
        if flask_app is not None:
            from opentelemetry.instrumentation.flask import FlaskInstrumentor
            FlaskInstrumentor().instrument_app(flask_app)
            # IMPORTANTE: além da auto-instrumentação (que gera spans/traces),
            # registramos métricas HTTP CUSTOMIZADAS com nomes DETERMINÍSTICOS.
            #
            # Por quê? A auto-instrumentação do OTel emite `http.server.duration`
            # que, ao passar pelo Prometheus, vira `http_server_duration_milliseconds`
            # (ou variações conforme a versão e a convenção semântica). Esse nome
            # é IMPREVISÍVEL entre versões. Como nossos alertas e o dashboard
            # dependem de um nome ESTÁVEL, criamos as nossas próprias métricas:
            #   - http_requests_total           (counter, com label http_response_status_code)
            #   - http_request_duration_seconds  (histogram, em segundos)
            # Esses nomes batem exatamente com o PromQL em 06-prometheus-rules.yaml.
            _register_flask_http_metrics(flask_app, service_name)

        # Requests (chamadas HTTP de saída — auth-service chama flag-service, etc.)
        try:
            from opentelemetry.instrumentation.requests import RequestsInstrumentor
            RequestsInstrumentor().instrument()
        except ImportError:
            pass

        # Psycopg2 (chamadas a PostgreSQL viram spans automáticos)
        try:
            from opentelemetry.instrumentation.psycopg2 import Psycopg2Instrumentor
            Psycopg2Instrumentor().instrument()
        except ImportError:
            pass

        # Boto3/SDK AWS (SQS, DynamoDB do analytics-service)
        try:
            from opentelemetry.instrumentation.botocore import BotocoreInstrumentor
            BotocoreInstrumentor().instrument()
        except ImportError:
            pass

        # Logging: injeta trace_id em todos os logs do Python
        try:
            from opentelemetry.instrumentation.logging import LoggingInstrumentor
            LoggingInstrumentor().instrument(set_logging_format=True)
        except ImportError:
            pass

        log.info(
            "OpenTelemetry inicializado",
            extra={
                "otel_service_name": service_name,
                "otel_endpoint": endpoint,
            },
        )

    except Exception as e:
        # FALHA do OTel JAMAIS pode derrubar a aplicação — log e segue.
        log.error("Falha ao inicializar OpenTelemetry (continuando sem): %s", e)


def _register_flask_http_metrics(flask_app, service_name: str) -> None:
    """
    Registra métricas HTTP customizadas via before_request/after_request do Flask.

    Cria duas métricas com nomes ESTÁVEIS (não dependem da convenção semântica
    do OTel, que varia entre versões):

      http_requests_total
        - tipo: counter
        - labels: service, http_request_method, http_route, http_response_status_code
        - uso: taxa de requisições (RPS) e taxa de erro 5xx

      http_request_duration_seconds
        - tipo: histogram (em SEGUNDOS)
        - labels: service, http_request_method, http_route
        - uso: latência p95 via histogram_quantile

    Esses nomes/labels são EXATAMENTE os que aparecem em:
      - gitops/base/observability/06-prometheus-rules.yaml
      - gitops/base/observability/07-grafana-dashboard.yaml
    """
    import time
    from flask import request
    from opentelemetry import metrics

    meter = metrics.get_meter("togglemaster.http", "1.0.0")

    # NOTA SOBRE NOMES: o exporter Prometheus adiciona AUTOMATICAMENTE o sufixo
    # `_total` a counters e `_bucket/_sum/_count` a histograms. Por isso o
    # counter é criado como "http_requests" (sem _total) — o Prometheus o
    # expõe como "http_requests_total". Se criássemos já com "_total", viraria
    # "http_requests_total_total" (bug conhecido). O histogram vira
    # "http_request_duration_seconds_bucket/_sum/_count".
    requests_total = meter.create_counter(
        name="http_requests",
        description="Total de requisições HTTP processadas",
        unit="1",
    )
    request_duration = meter.create_histogram(
        name="http_request_duration_seconds",
        description="Duração das requisições HTTP em segundos",
        unit="s",
    )

    @flask_app.before_request
    def _start_timer():
        request._otel_start_time = time.perf_counter()

    @flask_app.after_request
    def _record_metrics(response):
        try:
            elapsed = time.perf_counter() - getattr(
                request, "_otel_start_time", time.perf_counter()
            )
            # Usa a rota (regra de URL) e não o path bruto, para evitar
            # explosão de cardinalidade (ex: /flags/<name> em vez de /flags/abc).
            route = request.url_rule.rule if request.url_rule else "unmatched"
            base_attrs = {
                "service": service_name,
                "http_request_method": request.method,
                "http_route": route,
            }
            requests_total.add(
                1,
                {**base_attrs, "http_response_status_code": str(response.status_code)},
            )
            request_duration.record(elapsed, base_attrs)
        except Exception:
            # Nunca deixar a coleta de métrica quebrar a resposta ao usuário
            pass
        return response
